src/components/DistanceRequest.js

```diff
@@ -197,62 +186,69 @@ function DistanceRequest({transactionID, report, transaction, mapboxAccessToken,
-                                onPress={() => navigateToWaypointEditPage(index)}
-                                key={key}
-                            />
-                        );
-                    })}
-                </ScrollView>
-                {shouldShowGradient && (
-                    <LinearGradient
-                        style={[styles.pAbsolute, styles.b0, styles.l0, styles.r0, {height: halfMenuItemHeight}]}
-                        colors={[StyleUtils.getTransparentColor(theme.modalBackground), theme.modalBackground]}
-                    />
-                )}
-                {hasRouteError && (
-                    <DotIndicatorMessage
-                        style={[styles.mh5, styles.mv3]}
-                        messages={ErrorUtils.getLatestErrorField(transaction, 'route')}
-                        type="error"
-                    />
-                )}
-            </View>
+    const updateWaypoints = useCallback(
+        ({data}) => {
+            if (_.isEqual(waypointsList, data)) {
+                return;
+            }
+
+            const newWaypoints = {};
+            _.each(data, (waypoint, index) => {
+                const newWaypoint = lodashGet(waypoints, waypoint, {});
+                newWaypoints[`waypoint${index}`] = lodashIsEmpty(newWaypoint) ? null : newWaypoint;
+            });
+
+            setOptimisticWaypoints(newWaypoints);
+            // eslint-disable-next-line rulesdir/no-thenable-actions-in-views
+            Transaction.updateWaypoints(transactionID, newWaypoints).then(() => {
+                setOptimisticWaypoints(null);
+            });
+        },
+        [transactionID, waypoints, waypointsList],
+    );
+
+    const renderItem = ({item, drag, getIndex, isActive}) => {
+        const index = getIndex();
+        let descriptionKey = 'distance.waypointDescription.';
+        let waypointIcon;
+        if (index === 0) {
+            descriptionKey += 'start';
+            waypointIcon = Expensicons.DotIndicatorUnfilled;
+        } else if (index === lastWaypointIndex) {
+            descriptionKey += 'finish';
+            waypointIcon = Expensicons.Location;
+        } else {
+            descriptionKey += 'stop';
+            waypointIcon = Expensicons.DotIndicator;
+        }
+
+        return (
+            <MenuItemWithTopDescription
+                description={translate(descriptionKey)}
+                title={lodashGet(waypoints, [`waypoint${index}`, 'address'], '')}
+                icon={Expensicons.DragHandles}
+                iconFill={theme.icon}
+                secondaryIcon={waypointIcon}
+                secondaryIconFill={theme.icon}
+                shouldShowRightIcon
+                onPress={() => navigateToWaypointEditPage(index)}
+                onSecondaryInteraction={drag}
+                focused={isActive}
+                key={item}
+                disabled={isLoadingRoute}
+            />
+        );
+    };
+
+    const footer = (
```

Comment:
can you define a component for the footer and extract the definition out of `DistanceRequest` component?

Reply:
You can pass the component like this to `DraggableList`

```ts
function ListFooter() {
  return <Text>This is a footer.</Text>;
}

<DraggableList 
  // other props...
  ListFooterComponent={<ListFooter />}
/>
```